Orange/widgets/utils/filedialogs.py

```diff
@@ -85,3 +87,269 @@ def get_file_name(start_dir, start_filter, file_formats):
+
+    The mixin works by adding a `recent_path` setting storing a list of
+    instances of :obj:`RecentPath` (not pure strings). The widget can also
+    manipulate the settings directly when `add_path` and `last_path` do not
+    suffice.
+
+    If the widget has a simple combo box with file names, use
+    :obj:`RecentPathsWComboMixin`, which also manages the combo box.
+
+    Since this is a mixin, make sure to explicitly call its constructor by
+    `RecentPathsWidgetMixin.__init__(self)`.
+    """
+
+    #: list with search paths; overload to add, say, documentation data sets dir
+    SEARCH_PATHS = []
+
+    #: List[RecentPath]
+    recent_paths = Setting([])
+
+    _init_called = False
+
+    def __init__(self):
+        super().__init__()
+        self._init_called = True
+        self._relocate_recent_files()
+
+    def _check_init(self):
+        if not self._init_called:
+            raise RuntimeError("RecentPathsWidgetMixin.__init__ was not called")
+
+    def _search_paths(self):
+        basedir = self.workflowEnv().get("basedir", None)
+        if basedir is None:
+            return self.SEARCH_PATHS
+        return self.SEARCH_PATHS + [("basedir", basedir)]
+
+    def _relocate_recent_files(self):
+        self._check_init()
+        search_paths = self._search_paths()
+        rec = []
+        for recent in self.recent_paths:
+            resolved = recent.resolve(search_paths)
+            if resolved is not None:
+                rec.append(
+                    RecentPath.create(resolved.abspath, search_paths))
+            elif recent.search(search_paths) is not None:
+                rec.append(
+                    RecentPath.create(recent.search(search_paths), search_paths)
+                )
+        # change the list in-place for the case the widgets wraps this list
+        # in some model (untested!)
+        self.recent_paths[:] = rec
+
+    def workflowEnvChanged(self, key, value, oldvalue):
+        """
+        Handle changes of the working directory
+
+        The function is triggered by a signal from the canvas when the user
+        saves the schema.
+        """
+        if key == "basedir":
+            self._relocate_recent_files()
+
+    def add_path(self, filename):
+        """Add (or move) a file name to the top of recent paths"""
+        self._check_init()
+        recent = RecentPath.create(filename, self._search_paths())
+        if recent in self.recent_paths:
+            self.recent_paths.remove(recent)
+        self.recent_paths.insert(0, recent)
+
+    def select_file(self, n):
+        """Move the n-th file to the top of the list"""
+        recent = self.recent_paths[n]
+        del self.recent_paths[n]
+        self.recent_paths.insert(0, recent)
+
+    def last_path(self):
+        """Return the most recent absolute path or `None` if there is none"""
+        abspath = self.recent_paths and self.recent_paths[0].abspath
+        if abspath != "(none)":
+            return abspath
```

Comment:
if `self.recent_paths` is `[]` (the default value) this will return an `[]` not a `None`. Also the abspath of the recent item should never be a string "(none)".